torch_tvm/compiler.cpp

```diff
@@ -352,37 +358,37 @@ void TVMCompiler::run(Stack& stack) {
   // Parameters are managed by cached tvm_param_tensors.
   // They get deallocated when cache_ is deleted.
   std::vector<DLManagedTensorPtr> dl_tensor_list;
-  for (auto i = 0; i < cache_[spec].input_values.size(); ++i) {
-    auto* value = cache_[spec].input_values[i].first;
-    bool input_immutable = cache_[spec].input_values[i].second;
+  for (auto& input_value : cache_[spec].input_values) {
+    auto* value = input_value.first;
+    TVMGraphInputInfo& graph_input = input_value.second;
     if (!value_to_ivalue.count(value)) {
       auto optional_ivalue = toIValue(value);
       TORCH_INTERNAL_ASSERT(optional_ivalue.has_value());
       value_to_ivalue[value] = optional_ivalue.value();
     }
-    auto ivalue = value_to_ivalue.at(cache_[spec].input_values[i].first);
+    auto ivalue = value_to_ivalue.at(value);
     //auto tensor = ivalue.toTensor().to(at::kFloat);
     auto tensor = ivalue.toTensor();
     DLManagedTensor* dl_tensor;
     if (tensor.is_contiguous() &&
         torch_tvm::utils::isAligned(tensor.data_ptr(),
           tvm::runtime::kAllocAlignment)) {
       dl_tensor = at::toDLPack(tensor);
-    } else if (input_immutable) {
-      if (cache_[spec].tvm_param_tensors.find(value) ==
-        cache_[spec].tvm_param_tensors.end()) {
-        dl_tensor = torch_tvm::utils::allocAndCopyData(tensor);
-        cache_[spec].tvm_param_tensors[value] = DLManagedTensorPtr(dl_tensor);
+    } else if (graph_input.is_param) {
+      if (graph_input.tvm_tensor) {
+        dl_tensor = graph_input.tvm_tensor.get();
       } else {
-        dl_tensor = cache_[spec].tvm_param_tensors[value].get();
+        dl_tensor = torch_tvm::utils::allocAndCopyData(tensor);
+        graph_input.tvm_tensor = DLManagedTensorPtr(dl_tensor);
       }
     } else {
       dl_tensor =
         torch_tvm::utils::allocAndCopyData(tensor);
       dl_tensor_list.emplace_back(
           dl_tensor);
     }
-    cache_[spec].set_input(i, tvm::runtime::NDArray::FromDLPack(dl_tensor));
+    cache_[spec].set_input(graph_input.tvm_var_name,
```

Comment:
I'd keep a map of IValue* and int. And set input by position. Hash a string is slower than hash a pointer.

Reply:
I think the position of the nodes is known to runtime only. Internally it just does a search to find the index. I will see if it exposes indexed node names.

Reply:
Let me see if this is possible to do. Position of the node for a given variable name is known by the runtime. I am not sure if this is exposed, but I will look into it.

Reply:
OK. It's fine for this diff I think.